# JSON examples and exercise
****
+ get familiar with packages for dealing with JSON
+ study examples with JSON strings and files 
+ work on exercise to be completed and submitted 
****
+ reference: http://pandas.pydata.org/pandas-docs/stable/io.html#io-json-reader
+ data source: http://jsonstudio.com/resources/
****

In [ ]:
import pandas as pd

## imports for Python, Pandas

In [ ]:
import json
from pandas.io.json import json_normalize

## JSON example, with string

+ demonstrates creation of normalized dataframes (tables) from nested json string
+ source: http://pandas.pydata.org/pandas-docs/stable/io.html#normalization

In [ ]:
# define json string
data = [{'state': 'Florida', 
         'shortname': 'FL',
         'info': {'governor': 'Rick Scott'},
         'counties': [{'name': 'Dade', 'population': 12345},
                      {'name': 'Broward', 'population': 40000},
                      {'name': 'Palm Beach', 'population': 60000}]},
        {'state': 'Ohio',
         'shortname': 'OH',
         'info': {'governor': 'John Kasich'},
         'counties': [{'name': 'Summit', 'population': 1234},
                      {'name': 'Cuyahoga', 'population': 1337}]}]

In [ ]:
# use normalization to create tables from nested element
json_normalize(data, 'counties')

In [ ]:
# further populate tables created from nested element
json_normalize(data, 'counties', ['state', 'shortname', ['info', 'governor']])

****
## JSON example, with file

+ demonstrates reading in a json file as a string and as a table
+ uses small sample file containing data about projects funded by the World Bank 
+ data source: http://jsonstudio.com/resources/

In [ ]:
# load json as string
json.load((open('data/world_bank_projects_less.json')))

In [ ]:
# load as Pandas dataframe
sample_json_df = pd.read_json('data/world_bank_projects_less.json')
sample_json_df

****
## JSON exercise

Using data in file 'data/world_bank_projects.json' and the techniques demonstrated above,
1. Find the 10 countries with most projects
2. Find the top 10 major project themes (using column 'mjtheme_namecode')
3. In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.

In [1]:
# import packages needed
import pandas as pd
import json
from pandas.io.json import json_normalize

In [2]:
# load json, save as string
json = json.load((open('data/world_bank_projects.json')))

In [3]:
# normalize json into dataframe
json_df = json_normalize(json)
json_df.head()

_id.$oid approvalfy board_approval_month  \
0  52b213b38594d8a2be17c780       1999             November   
1  52b213b38594d8a2be17c781       2015             November   
2  52b213b38594d8a2be17c782       2014             November   
3  52b213b38594d8a2be17c783       2014              October   
4  52b213b38594d8a2be17c784       2014              October   

      boardapprovaldate                                 borrower  \
0  2013-11-12T00:00:00Z  FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA   
1  2013-11-04T00:00:00Z                    GOVERNMENT OF TUNISIA   
2  2013-11-01T00:00:00Z   MINISTRY OF FINANCE AND ECONOMIC DEVEL   
3  2013-10-31T00:00:00Z   MIN. OF PLANNING AND INT'L COOPERATION   
4  2013-10-31T00:00:00Z                      MINISTRY OF FINANCE   

            closingdate                              country_namecode  \
0  2018-07-07T00:00:00Z  Federal Democratic Republic of Ethiopia!$!ET   
1                   NaN                      Republic of Tunisia!$!TN   
2                   NaN                                   Tuvalu!$!TV   
3                   NaN                        Republic of Yemen!$!RY   
4  2019-04-30T00:00:00Z                       Kingdom of Lesotho!$!LS   

  countrycode                              countryname    countryshortname  \
0          ET  Federal Democratic Republic of Ethiopia            Ethiopia   
1          TN                      Republic of Tunisia             Tunisia   
2          TV                                   Tuvalu              Tuvalu   
3          RY                        Republic of Yemen  Yemen, Republic of   
4          LS                       Kingdom of Lesotho             Lesotho   

                         ...                         source  status  \
0                        ...                           IBRD  Active   
1                        ...                           IBRD  Active   
2                        ...                           IBRD  Active   
3                        ...                           IBRD  Active   
4                        ...                           IBRD  Active   

   supplementprojectflg                             theme1.Name  \
0                     N                       Education for all   
1                     N               Other economic management   
2                     Y                    Regional integration   
3                     N      Participation and civic engagement   
4                     N  Export development and competitiveness   

  theme1.Percent                                     theme_namecode  \
0            100      [{'code': '65', 'name': 'Education for all'}]   
1             30  [{'code': '24', 'name': 'Other economic manage...   
2             46  [{'code': '47', 'name': 'Regional integration'...   
3             50  [{'code': '57', 'name': 'Participation and civ...   
4             30  [{'code': '45', 'name': 'Export development an...   

     themecode   totalamt totalcommamt  \
0           65  130000000    130000000   
1        54,24          0      4700000   
2  52,81,25,47    6060000      6060000   
3        59,57          0      1500000   
4        41,45   13100000     13100000   

                                                 url  
0  http://www.worldbank.org/projects/P129828/ethi...  
1  http://www.worldbank.org/projects/P144674?lang=en  
2  http://www.worldbank.org/projects/P145310?lang=en  
3  http://www.worldbank.org/projects/P144665?lang=en  
4  http://www.worldbank.org/projects/P144933/seco...  

[5 rows x 55 columns]

In [4]:
# group by country and run count on number of entries
json_df.groupby('country_namecode')['country_namecode'].count().sort_values(ascending=False)

country_namecode
People's Republic of China!$!CN                  19
Republic of Indonesia!$!ID                       19
Socialist Republic of Vietnam!$!VN               17
Republic of India!$!IN                           16
Republic of Yemen!$!RY                           13
Nepal!$!NP                                       12
People's Republic of Bangladesh!$!BD             12
Kingdom of Morocco!$!MA                          12
Africa!$!3A                                      11
Republic of Mozambique!$!MZ                      11
Burkina Faso!$!BF                                 9
Federative Republic of Brazil!$!BR                9
Islamic Republic of Pakistan!$!PK                 9
United Republic of Tanzania!$!TZ                  8
Republic of Tajikistan!$!TJ                       8
Republic of Armenia!$!AM                          8
Kyrgyz Republic!$!KG                              7
Lao People's Democratic Republic!$!LA             7
Hashemite Kingdom of Jordan!$!JO               

****

## Country Project Count

After normalizing the data and taking a count of the number of project entries for each country, the top 10 countries are:
1. China - 19
2. Indonesia - 19
3. Vietnam - 17
4. India - 16
5. Yemen - 13
6. Nepal - 12
7. Bangladesh - 12
8. Morocco - 12
9. Africa - 11
10. Mozambique - 11

The one result that is suspect is the entry for "Africa". Below, we can look at rows with "Africa" as a 'countryname' to see what countries within Africa this entry is referring to.

In [5]:
# explore what entries are listed under "Africa"
json_df[json_df['countryname']=='Africa']

_id.$oid approvalfy board_approval_month  \
45   52b213b38594d8a2be17c7ad       2014            September   
46   52b213b38594d8a2be17c7ae       2014            September   
51   52b213b38594d8a2be17c7b3       2014            September   
58   52b213b38594d8a2be17c7ba       2014               August   
65   52b213b38594d8a2be17c7c1       2014               August   
99   52b213b38594d8a2be17c7e3       2013                 June   
167  52b213b38594d8a2be17c827       2013                  May   
184  52b213b38594d8a2be17c838       2013                  May   
287  52b213b38594d8a2be17c89f       2013                March   
353  52b213b38594d8a2be17c8e1       2013              January   
449  52b213b38594d8a2be17c941       2013              October   

        boardapprovaldate                       borrower  \
45   2013-09-12T00:00:00Z                         ECOWAS   
46   2013-09-10T00:00:00Z                 UGANDA-COMOROS   
51   2013-09-04T00:00:00Z               OSS, IUCN, CILSS   
58   2013-08-28T00:00:00Z                  BANK EXECUTED   
65   2013-08-06T00:00:00Z        BURUNDI,RWANDA,TANZANIA   
99   2013-06-28T00:00:00Z             GOVERNMENT OF MALI   
167  2013-05-30T00:00:00Z  MINISTRIES OF TELECOM AND ICT   
184  2013-05-21T00:00:00Z                       TANZANIA   
287  2013-03-14T00:00:00Z     MALAWI, MOZAMBIQUE, ZAMBIA   
353  2013-01-01T00:00:00Z          NILE BASIN INITIATIVE   
449  2012-10-02T00:00:00Z            GOVERNMENT OF NIGER   

              closingdate country_namecode countrycode countryname  \
45                    NaN      Africa!$!3A          3A      Africa   
46   2018-06-30T00:00:00Z      Africa!$!3A          3A      Africa   
51                    NaN      Africa!$!3A          3A      Africa   
58                    NaN      Africa!$!3A          3A      Africa   
65   2020-12-31T00:00:00Z      Africa!$!3A          3A      Africa   
99                    NaN      Africa!$!3A          3A      Africa   
167  2018-11-30T00:00:00Z      Africa!$!3A          3A      Africa   
184  2018-12-31T00:00:00Z      Africa!$!3A          3A      Africa   
287  2020-01-31T00:00:00Z      Africa!$!3A          3A      Africa   
353  2015-12-31T00:00:00Z      Africa!$!3A          3A      Africa   
449  2020-04-01T00:00:00Z      Africa!$!3A          3A      Africa   

    countryshortname                        ...                          \
45            Africa                        ...                           
46            Africa                        ...                           
51            Africa                        ...                           
58            Africa                        ...                           
65            Africa                        ...                           
99            Africa                        ...                           
167           Africa                        ...                           
184           Africa                        ...                           
287           Africa                        ...                           
353           Africa                        ...                           
449           Africa                        ...                           

    source  status  supplementprojectflg  \
45    IBRD  Active                     N   
46    IBRD  Active                     N   
51    IBRD  Active                     N   
58    IBRD  Active                     N   
65    IBRD  Active                     N   
99    IBRD  Active                     Y   
167   IBRD  Active                     N   
184   IBRD  Active                     N   
287   IBRD  Active                     N   
353   IBRD  Active                     N   
449   IBRD  Active                     N   

                                           theme1.Name theme1.Percent  \
45                           Health system performance            100   
46             Administrative and civil service reform             20   
51                           

****

## Country Project Count (cont.)

Under the 'borrower' column, it looks like there are a number of different groups that have projects, such as the Government of Mali, Tanzania, Niger, and others. If we revise our Top 10 to include this fact, then we have:

1. China - 19
2. Indonesia - 19
3. Vietnam - 17
4. India - 16
5. Yemen - 13
6. Nepal - 12
7. Bangladesh - 12
8. Morocco - 12
9. Mozambique - 11
10. Burkina Faso - 9

In [6]:
# check the project theme namecodes
themes = json_normalize(json, 'mjtheme_namecode')
themes

code                                          name
0       8                             Human development
1      11                                              
2       1                           Economic management
3       6         Social protection and risk management
4       5                         Trade and integration
5       2                      Public sector governance
6      11  Environment and natural resources management
7       6         Social protection and risk management
8       7                   Social dev/gender/inclusion
9       7                   Social dev/gender/inclusion
10      5                         Trade and integration
11      4      Financial and private sector development
12      6         Social protection and risk management
13      6                                              
14      2                      Public sector governance
15      4      Financial and private sector development
16     11  Environment and natural resources management
17      8                                              
18     10                             Rural development
19      7                                              
20      2                      Public sector governance
21      2                      Public sector governance
22      2                      Public sector governance
23     10                             Rural development
24      2                                              
25     10                             Rural development
26      6         Social protection and risk management
27      6         Social protection and risk management
28     11  Environment and natural resources management
29      4                                              
...   ...                                           ...
1469    8                             Human development
1470    9                             Urban development
1471    6         Social protection and risk management
1472    6         Social protection and risk management
1473    9                             Urban development
1474    2                      Public sector governance
1475    2                      Public sector governance
1476   10                             Rural development
1477   11                                              
1478    8                             Human development
1479    7                   Social dev/gender/inclusion
1480   11  Environment and natural resources management
1481    5                                              
1482    6         Social protection and risk management
1483    8                                              
1484    4      Financial and private sector development
1485    7                   Social dev/gender/inclusion
1486    8                             Human development
1487    5                         Trade and integration
1488    2                      Public sector governance
1489    8                             Human development
1490   10                             Rural development
1491    6                                              
1492   10                             Rural development
1493   10                             Rural development
1494   10                             Rural development
1495    9                             Urban development
1496    8                             Human development
1497    5                         Trade and integration
1498    4      Financial and private sector development

[1499 rows x 2 columns]

In [7]:
# notice that one of the entries is ''
themes['name'].unique()

array(['Human development', '', 'Economic management',
       'Social protection and risk management', 'Trade and integration',
       'Public sector governance',
       'Environment and natural resources management',
       'Social dev/gender/inclusion',
       'Financial and private sector development', 'Rural development',
       'Urban development', 'Rule of law'], dtype=object)

In [8]:
# create dict for code:name
theme_dict = {'1':'Economic management', 
              '2':'Public sector governance', 
              '3':'Rule of law', 
              '4':'Financial and private sector development', 
              '5':'Trade and integration', 
              '6':'Social protection and risk management', 
              '7':'Social dev/gender/inclusion', 
              '8':'Human development', 
              '9':'Urban development', 
              '10':'Rural development', 
              '11':'Environmental and natural resources management'}

In [9]:
# make 'name' column a duplicate of 'code', and then use .replace() based on dict
themes['name'] = themes['code']
themes['name'].replace(theme_dict, inplace=True)
themes

code                                            name
0       8                               Human development
1      11  Environmental and natural resources management
2       1                             Economic management
3       6           Social protection and risk management
4       5                           Trade and integration
5       2                        Public sector governance
6      11  Environmental and natural resources management
7       6           Social protection and risk management
8       7                     Social dev/gender/inclusion
9       7                     Social dev/gender/inclusion
10      5                           Trade and integration
11      4        Financial and private sector development
12      6           Social protection and risk management
13      6           Social protection and risk management
14      2                        Public sector governance
15      4        Financial and private sector development
16     11  Environmental and natural resources management
17      8                               Human development
18     10                               Rural development
19      7                     Social dev/gender/inclusion
20      2                        Public sector governance
21      2                        Public sector governance
22      2                        Public sector governance
23     10                               Rural development
24      2                        Public sector governance
25     10                               Rural development
26      6           Social protection and risk management
27      6           Social protection and risk management
28     11  Environmental and natural resources management
29      4        Financial and private sector development
...   ...                                             ...
1469    8                               Human development
1470    9                               Urban development
1471    6           Social protection and risk management
1472    6           Social protection and risk management
1473    9                               Urban development
1474    2                        Public sector governance
1475    2                        Public sector governance
1476   10                               Rural development
1477   11  Environmental and natural resources management
1478    8                               Human development
1479    7                     Social dev/gender/inclusion
1480   11  Environmental and natural resources management
1481    5                           Trade and integration
1482    6           Social protection and risk management
1483    8                               Human development
1484    4        Financial and private sector development
1485    7                     Social dev/gender/inclusion
1486    8                               Human development
1487    5                           Trade and integration
1488    2                        Public sector governance
1489    8                               Human development
1490   10                               Rural development
1491    6           Social protection and risk management
1492   10                               Rural development
1493   10                               Rural development
1494   10                               Rural development
1495    9                               Urban development
1496    8                               Human development
1497    5                           Trade and integration
1498    4        Financial and private sector development

[1499 rows x 2 columns]

In [10]:
# now can check groupby to see which projects have the highest count
themes.groupby('name')['name'].count().sort_values(ascending=False)

name
Environmental and natural resources management    250
Rural development                                 216
Human development                                 210
Public sector governance                          199
Social protection and risk management             168
Financial and private sector development          146
Social dev/gender/inclusion                       130
Trade and integration                              77
Urban development                                  50
Economic management                                38
Rule of law                                        15
Name: name, dtype: int64

**** 

## Project Theme Count

As it turns out, there are only 11 unique themes, which means the Top 10 will be all but one. They are:

1. Environmental and natural resources management - 250
2. Rural development - 216
3. Human development - 210
4. Public sector governance - 199
5. Social protection and risk management - 168
6. Financial and private sector development - 146
7. Social dev/gender/inclusion - 130
8. Trade and integration - 77
9. Urban development - 50
10. Economic management - 38